File: ASP.NET Web Forms Application Using Entity Framework Database First - V2/[C#]-ASP.NET Web Forms Application Using Entity Framework Database First - V2/C#/ContosoUniversity/BLL/SchoolBL.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using ContosoUniversity.DAL;

namespace ContosoUniversity.BLL
{
    public class SchoolBL : IDisposable
    {
        private ISchoolRepository schoolRepository;

        public SchoolBL()
        {
            this.schoolRepository = new SchoolRepository();
        }

        public SchoolBL(ISchoolRepository schoolRepository)
        {
            this.schoolRepository = schoolRepository;
        }

        public IEnumerable<Department> GetDepartments()
        {
            return schoolRepository.GetDepartments();
        }

        public IEnumerable<Department> GetDepartments(string sortExpression)
        {
            return schoolRepository.GetDepartments(sortExpression);
        }

        public IEnumerable<Department> GetDepartmentsByName(string sortExpression, string nameSearchString)
        {
            return schoolRepository.GetDepartmentsByName(sortExpression, nameSearchString);
        }
        
        public void InsertDepartment(Department department)
        {
            ValidateOneAdministratorAssignmentPerInstructor(department);
            try
            {
                schoolRepository.InsertDepartment(department);
            }
            catch (Exception ex)
            {
                //Include catch blocks for specific exceptions first, 
                //and handle or log the error as appropriate in each. 
                //Include a generic catch block like this one last. 
                throw ex;
            }
        }

        public void DeleteDepartment(Department department)
        {
            try
            {
                schoolRepository.DeleteDepartment(department);
            }
            catch (Exception ex)
            {
                //Include catch blocks for specific exceptions first, 
                //and handle or log the error as appropriate in each. 
                //Include a generic catch block like this one last. 
                throw ex;
            }
        }

        public void UpdateDepartment(Department department, Department origDepartment)
        {
            ValidateOneAdministratorAssignmentPerInstructor(department);
            try
            {
                schoolRepository.UpdateDepartment(department, origDepartment);
            }
            catch (Exception ex)
            {
                //Include catch blocks for specific exceptions first, 
                //and handle or log the error as appropriate in each. 
                //Include a generic catch block like this one last. 
                throw ex;
            }

        }

        public IEnumerable<InstructorName> GetInstructorNames()
        {
            return schoolRepository.GetInstructorNames();
        }

        public IEnumerable<OfficeAssignment> GetOfficeAssignments(string sortExpression)
        {
            if (string.IsNullOrEmpty(sortExpression)) sortExpression = "Person.LastName";
            return schoolRepository.GetOfficeAssignments(sortExpression);
        }

        public void InsertOfficeAssignment(OfficeAssignment officeAssignment)
        {
            try
            {
                schoolRepository.InsertOfficeAssignment(officeAssignment);
            }
            catch (Exception ex)
            {
                //Include catch blocks for specific exceptions first, 
                //and handle or log the error as appropriate in each. 
                //Include a generic catch block like this one last. 
                throw ex;
            }
        }

        public void DeleteOfficeAssignment(OfficeAssignment officeAssignment)
        {
            try
            {
                schoolRepository.DeleteOfficeAssignment(officeAssignment);
            }
            catch (Exception ex)
            {
                //Include catch blocks for specific exceptions first, 
                //and handle or log the error as appropriate in each. 
                //Include a generic catch block like this one last. 
                throw ex;
            }
        }

        public void UpdateOfficeAssignment(OfficeAssignment officeAssignment, OfficeAssignment origOfficeAssignment)
        {
            try
            {
                schoolRepository.UpdateOfficeAssignment(officeAssignment, origOfficeAssignment);
            }
            catch (Exception ex)
            {
                //Include catch blocks for specific exceptions first, 
                //and handle or log the error as appropriate in each. 
                //Include a generic catch block like this one last. 
                throw ex;
            }
        }
        
        private void ValidateOneAdministratorAssignmentPerInstructor(Department department)
        {
            if (department.Administrator != null)
            {
                var duplicateDepartment = schoolRepository.GetDepartmentsByAdministrator(department.Administrator.GetValueOrDefault()).FirstOrDefault();
                if (duplicateDepartment != null && duplicateDepartment.DepartmentID != department.DepartmentID)
                {
                    throw new DuplicateAdministratorException(String.Format(
                        "Instructor {0} {1} is already administrator of the {2} department.",
                        duplicateDepartment.Person.FirstMidName,
                        duplicateDepartment.Person.LastName,
                        duplicateDepartment.Name));
                }
            }
        }

        private bool disposedValue = false;

        protected virtual void Dispose(bool disposing)
        {
            if (!this.disposedValue)
            {
                if (disposing)
                {
                    schoolRepository.Dispose();
                }
            }
            this.disposedValue = true;
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

    }
}

File: ASP.NET Web Forms Application Using Entity Framework Database First - V2/[VB]-ASP.NET Web Forms Application Using Entity Framework Database First - V2/VB/ContosoUniversity/BLL/SchoolBL.vb
Public Class SchoolBL
    Private schoolRepository As ISchoolRepository

    Public Sub New()
        Me.schoolRepository = New SchoolRepository()
    End Sub

    Public Sub New(ByVal schoolRepository As ISchoolRepository)
        Me.schoolRepository = schoolRepository
    End Sub

    Public Function GetDepartments() As IEnumerable(Of Department)
        Return schoolRepository.GetDepartments()
    End Function

    Public Function GetDepartments(ByVal sortExpression As String) As IEnumerable(Of Department)
        Return schoolRepository.GetDepartments(sortExpression)
    End Function

    Public Function GetDepartmentsByName(ByVal sortExpression As String, ByVal nameSearchString As String) As IEnumerable(Of Department)
        Return schoolRepository.GetDepartmentsByName(sortExpression, nameSearchString)
    End Function

    Public Sub InsertDepartment(ByVal department As Department)
        ValidateOneAdministratorAssignmentPerInstructor(department)
        Try
            schoolRepository.InsertDepartment(department)
        Catch ex As Exception
            'Include catch blocks for specific exceptions first, 
            'and handle or log the error as appropriate in each. 
            'Include a generic catch block like this one last. 
            Throw ex
        End Try
    End Sub

    Public Sub DeleteDepartment(ByVal department As Department)
        Try
            schoolRepository.DeleteDepartment(department)
        Catch ex As Exception
            'Include catch blocks for specific exceptions first, 
            'and handle or log the error as appropriate in each. 
            'Include a generic catch block like this one last. 
            Throw ex
        End Try
    End Sub

    Public Sub UpdateDepartment(ByVal department As Department, ByVal origDepartment As Department)
        ValidateOneAdministratorAssignmentPerInstructor(department)
        Try
            schoolRepository.UpdateDepartment(department, origDepartment)
        Catch ex As Exception
            'Include catch blocks for specific exceptions first, 
            'and handle or log the error as appropriate in each. 
            'Include a generic catch block like this one last. 
            Throw ex
        End Try

    End Sub

    Public Function GetInstructorNames() As IEnumerable(Of InstructorName)
        Return schoolRepository.GetInstructorNames()
    End Function

    Public Function GetOfficeAssignments(ByVal sortExpression As String) As IEnumerable(Of OfficeAssignment)
        If String.IsNullOrEmpty(sortExpression) Then
            sortExpression = "Person.LastName"
        End If
        Return schoolRepository.GetOfficeAssignments(sortExpression)
    End Function

    Public Sub InsertOfficeAssignment(ByVal officeAssignment As OfficeAssignment)
        Try
            schoolRepository.InsertOfficeAssignment(officeAssignment)
        Catch ex As Exception
            'Include catch blocks for specific exceptions first, 
            'and handle or log the error as appropriate in each. 
            'Include a generic catch block like this one last. 
            Throw ex
        End Try
    End Sub

    Public Sub DeleteOfficeAssignment(ByVal officeAssignment As OfficeAssignment)
        Try
            schoolRepository.DeleteOfficeAssignment(officeAssignment)
        Catch ex As Exception
            'Include catch blocks for specific exceptions first, 
            'and handle or log the error as appropriate in each. 
            'Include a generic catch block like this one last. 
            Throw ex
        End Try
    End Sub

    Public Sub UpdateOfficeAssignment(ByVal officeAssignment As OfficeAssignment, ByVal origOfficeAssignment As OfficeAssignment)
        Try
            schoolRepository.UpdateOfficeAssignment(officeAssignment, origOfficeAssignment)
        Catch ex As Exception
            'Include catch blocks for specific exceptions first, 
            'and handle or log the error as appropriate in each. 
            'Include a generic catch block like this one last. 
            Throw ex
        End Try
    End Sub

    Private Sub ValidateOneAdministratorAssignmentPerInstructor(ByVal department As Department)
        If department.Administrator IsNot Nothing Then
            Dim duplicateDepartment = schoolRepository.GetDepartmentsByAdministrator(department.Administrator.GetValueOrDefault()).FirstOrDefault()
            If duplicateDepartment IsNot Nothing AndAlso duplicateDepartment.DepartmentID <> department.DepartmentID Then
                Throw New DuplicateAdministratorException(
                    String.Format("Instructor {0} {1} is already administrator of the {2} department.",
                                  duplicateDepartment.Person.FirstMidName,
                                  duplicateDepartment.Person.LastName,
                                  duplicateDepartment.Name))
            End If
        End If
    End Sub

    Private disposedValue As Boolean = False

    Protected Overridable Sub Dispose(ByVal disposing As Boolean)
        If Not Me.disposedValue Then
            If disposing Then
                schoolRepository.Dispose()
            End If
        End If
        Me.disposedValue = True
    End Sub

    Public Sub Dispose()
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub

End Class
